Dockerfile

```diff
@@ -0,0 +1,49 @@
+#########################################################
+# Build the sources and provide the result in a multi stage
+# docker container. The alpine build image has to match
+# the alpine image in the referencing runtime container.
+#########################################################
+FROM golang:1.10.1-alpine3.7 AS build-env
+
+# Set all env variables needed for go
+ENV GOBIN /go/bin
+ENV GOPATH /go
+
+# We need so that dep can fetch it's dependencies
+RUN apk --no-cache add git
+
+# Directory in workspace
+RUN mkdir -p "/go/src/github.com/Peripli/service-manager"
+COPY . "/go/src/github.com/Peripli/service-manager"
+WORKDIR "/go/src/github.com/Peripli/service-manager"
+
+# Install dep, dependencies and build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go get github.com/golang/dep/cmd/dep && \
```

Comment:
I think it will be better to copy the whole source code in one step and then do in one step dep ensure -vendor-only and build. Otherwise there will be 2 copy steps, 1 for dependencies and another one for the build.

Reply:
It's not better to copy the whole source because it will always copy the whole source and dep files in case of changes in the files. If you copy first the toml and lock it will do a layer cache on them and it will not copy them unless there is change (and it will not do dep build unless dep files have changed). 

if you have dep build and source build in one RUN instruction it will always do both dep ensure and source build if any file (even though it might not be dep related) has changed.

The idea is that this is an example case when having two layers is actually better than having one layer :P. See the example here 

```
ADD Gopkg.toml Gopkg.toml
ADD Gopkg.lock Gopkg.lock

RUN dep ensure --vendor-only

ADD <sources and stuff>

RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o main src/main.go
```

Here are some reads I found : 
https://blog.hasura.io/the-ultimate-guide-to-writing-dockerfiles-for-go-web-apps-336efad7012c 

https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#use-multi-stage-builds

https://medium.com/@pierreprinetti/the-go-dockerfile-d5d43af9ee3c

Reply:
Just something else I thought of which maybe I didn't make clear in the previous comment - its actually the COPY and ADD instructions  the nullify the caching for the RUN instructions. That's why we need to have seperate copying and seperate runs